scipy/fftpack/realtransforms.py

```diff
@@ -203,71 +262,31 @@ def _dct(x, type, n=None, axis=-1, overwrite_x=False, normalize=None):
     x : array-like
         input array.
     n : int, optional
-        Length of the transform.
+        Length of the transform.  If ``n < x.shape[axis]``, `x` is
+        truncated.  If ``n > x.shape[axis]``, `x` is zero-padded. The
+        default results in ``n = x.shape[axis]``.
     axis : int, optional
-        Axis along which the dct is computed. (default=-1)
+        Axis along which the dct is computed; the default is over the
+        last axis (i.e., ``axis=-1``).
     overwrite_x : bool, optional
-        If True the contents of x can be destroyed. (default=False)
+        If True, the contents of `x` can be destroyed; the default is False.
 
     Returns
     -------
-    z : real ndarray
+    z : ndarray
 
     """
-    tmp = np.asarray(x)
-    if not np.isrealobj(tmp):
-        raise TypeError("1st argument must be real sequence")
-
-    if n is None:
-        n = tmp.shape[axis]
-    else:
-        raise NotImplementedError("Padding/truncating not yet implemented")
-
-    if tmp.dtype == np.double:
-        if type == 1:
-            f = _fftpack.ddct1
-        elif type == 2:
-            f = _fftpack.ddct2
-        elif type == 3:
-            f = _fftpack.ddct3
-        else:
-            raise ValueError("Type %d not understood" % type)
-    elif tmp.dtype == np.float32:
-        if type == 1:
-            f = _fftpack.dct1
-        elif type == 2:
-            f = _fftpack.dct2
-        elif type == 3:
-            f = _fftpack.dct3
-        else:
-            raise ValueError("Type %d not understood" % type)
-    else:
-        raise ValueError("dtype %s not supported" % tmp.dtype)
-
-    if normalize:
-        if normalize == "ortho":
-            nm = 1
-        else:
-            raise ValueError("Unknown normalize mode %s" % normalize)
-    else:
-        nm = 0
-
+    x0, n, copy_made = __fix_shape(x, n, axis, 'DCT')
     if type == 1 and n < 2:
         raise ValueError("DCT-I is not defined for size < 2")
+    overwrite_x = overwrite_x or copy_made
+    nm = _get_norm_mode(normalize)
+    if x0.dtype == complex:
```

Comment:
There is a little glitch with handling non-native complex types: on a 64-bit linux system: `dct(np.arange(7, dtype=np.complex64))` raises `ValueError: dtype complex64 not supported`. At the same time transforming `.real` and `.imag ` separately (just like it's done below) works fine.